### ISSUE TITLE
ARC cache hit size worse after update to 0.7.1

### ISSUE THREAD

**Issue by Lalufu** [zfsonlinux/zfs#6521]
### System information

Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       |  Fedora
Distribution Version    |  25
Linux Kernel                 |  4.11.12-200
Architecture                 |  x86_64
ZFS Version                  |  0.7.1
SPL Version                  |  0.7.1

### Describe the problem you're observing
I've been running monitroing statistics on ARC size/distribution and cache hit rate for a long time on this machine. The hardware and load hasn't changed really before and after the update to 0.7.1.

The values graphed (from /proc/spl/kstat/zfs/arcstats) are:

- Maximum size [c_max]
- Current size [size]
- Target size [c]
- MRU size [p]
- MFU size [c-p]
![zfs-size](https://user-images.githubusercontent.com/5759461/29381062-493fdd92-82c8-11e7-8e6b-d0cc18860258.png)

- Hit rate [100*(hits/(hits+misses))]
![zfs-cache](https://user-images.githubusercontent.com/5759461/29381112-79dbceca-82c8-11e7-986c-97c2d6173a84.png)

The update from 0.6.5.11 to 0.7.1 happened near the end of week 32. The large blue spike in the first image marks the reboot after the 0.7.1 update.

As mentioned the general load on the machine hasn't changed after the update. No options are set for the ZFS or SPL modules. From the graphs two things can be seen:

- Less memory is allocated to ARC, especially MRU is almost gone
- The cache hit rate has dropped from being close to 100% (with some dips, mostly during weekly scrub runs) to hovering between 50 and 60%.

Question: is this a real drop, or has the meaning of the graphed values changed significantly in 0.7.1?
If it's a real drop, what could be the cause?

The machine is not generally memory constrained:

```
[sun ethan ~ :) 4]$ free; arcstat.py 
              total        used        free      shared  buff/cache   available
Mem:       16346100     4929728      735008       26520    10681364     9493668
Swap:       2046972         512     2046460
    time  read  miss  miss%  dmis  dm%  pmis  pm%  mmis  mm%  arcsz     c  
19:23:16    87     0      0     0    0     0    0     0    0   991M  1023M  
```

**Comment by behlendorf** [comment]
Quite possibly related to #6518 and #6453 where we also suspect a significant decrease in the MRU hit rate.  This may also be related to #6506 where there's some changes to the observed ARC behavior.  With the addition of the compressed ARC and ABD features there was a huge amount of change in this area.  We'll definitely dig in to this to determine what's going on, but a decrease in the MRU hit rate wasn't intended.

**Comment by dbavatar** [comment]
@Lalufu What does your /proc/spl/kstat/zfs/arcstats look like? I bet the distribution of bytes in there has heavily tilted leaving little room for data mru.

**Comment by jstenback** [comment]
Looking back at my graphs I see a similar thing starting when I first switched to a 0.7.0 rc build, and I also noticed this:

![image](https://user-images.githubusercontent.com/509034/29396756-b799c162-82ce-11e7-90a2-d509da8d0e8f.png)

Unclear to me whether that's the same problem or not, but hoping it'll be another useful data point.

**Comment by devZer0** [comment]
I have also observed some ARC weirdness i cannot explain . This way ARC usage also does not meet my expectation. Something weird is going on here or my expectation from ARC/L2ARC is wrong or the workload "too exotic".

I have a large ZFS fileserver where i run nightly rsync backup from >100 other machines.

If i freshly boot the server, after backing up the largest client with about 1Mio files, ARC grows to 2,3Gb with primarycache=metadata. 

ARC even grows if i set primarycache=none. How can this be? 

2,3GB for about 1Mio Files is more than 2k of metadata for each file.   
What 2k of metadata does a ZFS hosted file have? How does such an file's ARC entry compose?

As i see that the ARC demand for metadata is such high, for testing i added L2ARC to the system and set secondarycache=metadata (primarycache=all) and for my curiosity,I don't see real usage/benefit.

After complete rsync backup run with tens of millions of files, it`s just at 3,52G, which is slightly larger than ARC size of 2,3GB for the single client tested. Based on that single run i would have expeced much much more metadata to go into L2ARC.

On repeated rsync runs, even with primarycache=all,secondarycache=all  i still see an enormous amount of IOPs hitting the disks, even if there is nothing to backup and rsync only walks the directory trees lstat()'ing files to compare if something changed.

Maybe somebody has a clue if this is a bug, a design issue or if there are tunables which makes metadata caching more effective...

**Comment by bunder2015** [comment]
I think I might be affected by this as well.  My server has been running the same build since around the time ABD landed, and I just noticed that my MRU is basically nil.  ARC usage also seems to have slumped, it did peak out around 16GB (as per zfs_arc_max) a day or two after the last boot (~96 days ago), but has slowly been dropping off, it's around 10GB now.  L2 also shows a similar gradual downward sloping.

**Comment by behlendorf** [comment]
The issue described here may be addressed by PR #6635.  The root cause was a API change in Linux 4.8 kernel to the return value of the `global_page_state()` function.  This change resulted in the ARC incorrectly determining the amount free space on the system.  The https://github.com/zfsonlinux/zfs/pull/6635#issuecomment-329346079 shows the change in ARC behavior, specifically mfu size, before and after the fix.

We're rolling this fix in to the 0.7.2 tag which should be released soon.  We also plan to continue investigating any strange ARC behavior which may have been caused by ABD.  Please let us know if after updating to 0.7.2 (or cherry picking this fix) you still observe odd behavior.

**Comment by bunder2015** [comment]
Thanks, let me pull git head and try again.  Due to the nature of ARC this might take me a week to fully test.

**Comment by bunder2015** [comment]
I think my ARC is good again.  Will keep watching long-term. 

![a](https://i.imgur.com/JMzsuT2.png)

**Comment by bunder2015** [comment]
Maybe I spoke too soon.

![a](https://i.imgur.com/4tr0ar4.png)

**Comment by behlendorf** [comment]
@bunder2015 that is a sharp change in how the cache is being utilized.  Did the workload change significantly on the 5th?

**Comment by bunder2015** [comment]
Not that I know of.  It's mostly a low-use home NAS + LAMP setup, I actually wasn't home much on Fri-Sun either.  The 5th would have been Thursday, I don't remember doing anything major that day as well.

**Comment by bunder2015** [comment]
tuesday
![a](https://i.imgur.com/gB8Udpq.png)
wednesday
![b](https://i.imgur.com/Z1qr92r.png)

I'm going to assume this is normal.  I have one cron job at 6am that uses wget to mirror another website, and for some reason wget uses a large amount of "linux cache" (despite the fact that I don't have any non-ZFS filesystems) which might account for the drops, although I do find it strange that I can't seem to keep the ARC full.

**Comment by mailinglists35** [comment]
you can prevent cron rsync/wget/ etc storming the cache with [eatmydata](https://packages.debian.org/sid/eatmydata) (for writing) and [nocache](https://packages.debian.org/jessie/nocache) (for reading)

**Comment by behlendorf** [comment]
Closing, resolved in 0.7.2.
